[rgw]: Adding support for SSE-S3 in GET and PUT path

priyagpssehgal opened the pull request:
Introducing SSE-S3 RGW options and using them in KMS code.
Introducing the support for SSE-S3 in GET and PUT path.
In the PUT path, we check if BucketEncryption is enabled for the bucket,
which received PUT call. If yes, we determine if the encryption type is
AES256 (i.e., SSE-S3), then we fetch the KEK-ID from the bucket x-attrs
and use it to encrypt the data key. Thereafter, we call the 
generate-data-key. We store the KEK-ID and the wrapped data-key in the 
object x-attrs.

In the GET path, we simply pull out the KEK-ID from the object x-attr and 
decrypt the object.
NOTE: We only support Hashicorp Vault and the SSE-S3 KMS backend.

Signed-off-by: Priya Sehgal <priya.sehgal@flipkart.com>

rahuldevp committed:
rgw: Introduce BucketEncryption APIs to support SSE-S3 feature

This patch introduces support for 3 new BucketEncryption APIs which are listed
below and are helpful in supporting AWS SSE-S3 encryption mode.
PutBucketEncryption: https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutBucketEncryption.html
GetBucketEncryption: https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketEncryption.html
DeleteBucketEncryption: https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteBucketEncryption.html

The user provided parameters are parsed and stored in the bucket's extended
attributes RGW_ATTR_BUCKET_ENCRYPTION and
RGW_ATTR_BUCKET_ENCRYPTION_SSE_S3_KEY_ID.

Signed-off-by: Rahul Dev Parashar rahul.dev@flipkart.com

mdw-at-linuxbox committed:
Introducing SSE-S3 RGW options and using them in KMS code

Added SSE-S3 options for the Vault as the backend.

Signed-off-by: Priya Sehgal <priya.sehgal@flipkart.com>

Commit:
Adding support for SSE-S3 in GET and PUT path.

This patch introduces the support for SSE-S3 in GET anf PUT path.
In the PUT path, we chec if BucketEncryption is enabled for the bucket,
which received PUT call. If yes, we detemine if the encryption type is
AES256 (i.e., SSE-S3), then we fetch the KEK-ID from the bucket X-attrs
and use it to encrypt the data key. Thereafter, we call the
generate-data-key. We store the KEK-ID and the wrapped data-key in the
object xattrs.
In the GET path, we simply pull out the KEK-ID from the object xattr and
decrypt the object.
NOTE: We only support Hashicorp Vault and the SSE-S3 KMS backend.

Signed-off-by: Priya Sehgal <priya.sehgal@flipkart.com>

Commit:
Adding support for generating KEK from the Vault

Generating the KEK based on bucket owner UID in PutBucketEncryption.

Signed-off-by: Priya Sehgal <priya.sehgal@flipkart.com>

rahuldevp commented on src/rgw/rgw_crypt.cc:
```diff
@@ -1090,6 +1092,78 @@ int rgw_s3_prepare_encrypt(struct req_state* s,
     }
 
+    /* Checking bucket attributes if SSE is enabled. Currently only supporting SSE-S3 */
+    rgw::sal::Attrs buck_attrs(s->bucket_attrs);
+    auto aiter = buck_attrs.find(RGW_ATTR_BUCKET_ENCRYPTION);
+    if (aiter != buck_attrs.end()) {
```
On my PR in the Get call I received comments from Casey that this should be rewritten. (https://github.com/ceph/ceph/pull/42222#discussion_r665545701)
A similar comment might come here. So, we can rewrite this as:
`auto aiter = s->bucket_attrs.find(RGW_ATTR_BUCKET_ENCRYPTION_POLICY);
  if (aiter == s->bucket_attrs.end()) {`
  Same can be done for other references if any where get attribute is done.

rahuldevp commented on src/rgw/rgw_crypt.cc:
```diff
@@ -1090,6 +1092,78 @@ int rgw_s3_prepare_encrypt(struct req_state* s,
+    rgw::sal::Attrs buck_attrs(s->bucket_attrs);
+    auto aiter = buck_attrs.find(RGW_ATTR_BUCKET_ENCRYPTION);
+    if (aiter != buck_attrs.end()) {
+      ldpp_dout(s, 5) << "Found RGW_ATTR_BUCKET_ENCRYPTION attr bucket_name: "
```
Found -> found
Usually the small case is followed in the codebase unless we put it as Error or Notice.

rahuldevp committed:
rgw: Introduce BucketEncryption APIs to support SSE-S3 feature  (#4)

* rgw: Introduce BucketEncryption APIs to support SSE-S3 feature

This patch introduces support for 3 new BucketEncryption APIs which are listed
below and are helpful in supporting AWS SSE-S3 encryption mode.
PutBucketEncryption: https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutBucketEncryption.html
GetBucketEncryption: https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketEncryption.html
DeleteBucketEncryption: https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteBucketEncryption.html

The user provided parameters are parsed and stored in the bucket's extended
attributes RGW_ATTR_BUCKET_ENCRYPTION and
RGW_ATTR_BUCKET_ENCRYPTION_SSE_S3_KEY_ID.
Documentation for BucketEncryption APIs is added.
For new IAM policies details are added in respective test case file.

Signed-off-by: Rahul Dev Parashar <rahul.dev@flipkart.com>

Commit:
Merge branch 'aws_sse_s3' of https://github.com/flipkart-incubator/ceph into aws_sse_s3_priya
Merging modified x-attrs of bucket encryption with GET code.

Signed-off-by: Priya Sehgal <priya.sehgal@flipkart.com>
